arch/src/aarch64/gic.rs

```diff
@@ -21,104 +18,114 @@ pub enum Error {
-
-    /* We need to tell the kernel how many irqs to support with this vgic.
-    See the `layout` module for details.
-    */
-    let nr_irqs: u32 = super::layout::IRQ_MAX - super::layout::IRQ_BASE + 1;
-    let nr_irqs_ptr = &nr_irqs as *const u32;
-    let nr_irqs_attr = kvm_bindings::kvm_device_attr {
-        group: kvm_bindings::KVM_DEV_ARM_VGIC_GRP_NR_IRQS,
-        attr: 0,
-        addr: nr_irqs_ptr as u64,
-        flags: 0,
-    };
-    vgic_fd
-        .set_device_attr(&nr_irqs_attr)
-        .map_err(Error::SetDeviceAttribute)?;
-
-    /* Finalize the GIC.
-         See https://code.woboq.org/linux/linux/virt/kvm/arm/vgic/vgic-kvm-device.c.html#211.
-    */
-    let init_gic_attr = kvm_bindings::kvm_device_attr {
-        group: kvm_bindings::KVM_DEV_ARM_VGIC_GRP_CTRL,
-        attr: u64::from(kvm_bindings::KVM_DEV_ARM_VGIC_CTRL_INIT),
-        addr: 0,
-        flags: 0,
-    };
-    vgic_fd
-        .set_device_attr(&init_gic_attr)
-        .map_err(Error::SetDeviceAttribute)?;
-
-    Ok(vgic_fd)
-}
+/// Trait for GIC devices.
+pub trait GICDevice {
+    /// Returns the GIC version of the device
+    fn version(&self) -> u32;
 
-// Auxiliary functions for getting addresses and size of where the distributor and redistributor
-// are placed.
-/// Get the address of the GIC distributor.
-pub fn get_dist_addr() -> u64 {
-    super::layout::MAPPED_IO_START - KVM_VGIC_V3_DIST_SIZE
-}
+    /// Returns the file descriptor of the GIC device
+    fn device_fd(&self) -> &DeviceFd;
 
-/// Get the size of the GIC distributor.
-pub fn get_dist_size() -> u64 {
-    KVM_VGIC_V3_DIST_SIZE
-}
+    /// Returns an array with GIC device properties
+    fn device_properties(&self) -> &[u64];
+
+    /// Returns the fdt compatibility property of the device
+    fn fdt_compatibility(&self) -> &str;
+
+    /// Returns the maint_irq fdt property of the device
+    fn fdt_maint_irq(&self) -> u32;
+
+    /// Initialize a GIC device
+    fn init_device(vm: &VmFd, version: u32) -> Result<DeviceFd>
+    where
+        Self: Sized,
+    {
+        let mut gic_device = kvm_bindings::kvm_create_device {
+            type_: version,
+            fd: 0,
+            flags: 0,
+        };
+
+        let vgic_fd = vm.create_device(&mut gic_device).map_err(Error::CreateGIC);
 
-/// Get the address of the GIC redistributors.
-pub fn get_redists_addr(vcpu_count: u64) -> u64 {
-    get_dist_addr() - get_redists_size(vcpu_count)
+        vgic_fd
+    }
+
+    /// Set a GIC device attribute
+    fn set_device_attribute(
+        fd: &DeviceFd,
+        group: u32,
+        attr: u64,
+        addr: u64,
+        flags: u32,
+    ) -> Result<()>
+    where
+        Self: Sized,
+    {
+        let attr = kvm_bindings::kvm_device_attr {
+            group: group,
+            attr: attr,
+            addr: addr,
+            flags: flags,
+        };
+        fd.set_device_attr(&attr)
+            .map_err(Error::SetDeviceAttribute)?;
+
+        Ok(())
+    }
```

Comment:
This method seems to only forward the parameters to `fd.set_device_attr`. I don't know if it brings a lot of value.